### PR TITLE
UI: Catch 500 error on token endpoint and revert to legacy tokens

### LINKED ISSUE #4871: Running the UI in a 1.4 client against a 1.3 cluster gives 500 error at the ACLs pages
When running a 1.4 UI client against a 1.3 cluster of servers, the new ACL API endpoints give 500 errors, therefore the UI shows a 500 error page when navigating to the ACLs area

Also see:

https://github.com/hashicorp/consul/issues/4839#issuecomment-433541999

@rboyer

### PR DESCRIPTION
**Pull request by johncowen**
In some circumstances a consul 1.4 client could be running in an
un-upgraded 1.3 or lower cluster. Currently this gives a 500 error on
the new ACL token endpoint. Here we catch this specific 500 error/message
and set the users AccessorID to `null`. Elsewhere in the frontend we use
this fact (AccessorID being `null`) to decide whether to present the
legacy or the new ACL UI to the user.

Fixes #4871

### CODE REVIEW

File: ui-v2/app/routes/dc/acls/index.js [hunk 1]
```diff
@@ -7,14 +7,26 @@ import WithAclActions from 'consul-ui/mixins/acl/with-actions';
...
   },
   beforeModel: function(transition) {
-    return this.replaceWith('dc.acls.tokens');
+    return get(this, 'settings')
+      .findBySlug('token')
+      .then(token => {
+        // If you don't have a token set or you have a
+        // token set with AccessorID set to null (legacy mode)
+        // then rewrite to the new acls
+        if (!token || get(token, 'AccessorID') !== null) {
```
Comment (DingoEatingFuzz) [inline ui-v2/app/routes/dc/acls/index.js]: Shouldn't that be `get(token, 'AccessorID') === null`? Right now the code doesn't match the comment.

Comment (DingoEatingFuzz) [reply ui-v2/app/routes/dc/acls/index.js]: Just read the code below, looks like it's the comment that's wrong, not the code.

File: ui-v2/app/services/repository/token.js [hunk 2]
```diff
@@ -20,10 +21,22 @@ export default Service.extend({
...
-    });
+    return get(this, 'store')
+      .self(this.getModelName(), {
+        secret: secret,
+        dc: dc,
+      })
+      .catch(e => {
+        // If we get this 500 RPC error, it means we are a legacy ACL cluster
+        // set AccessorID to null - which for the frontend means legacy mode
+        if (e.errors[0].detail.indexOf(UNKNOWN_METHOD_ERROR) === 0) {
```
Comment (DingoEatingFuzz) [inline ui-v2/app/services/repository/token.js]: I guess this works. It would be nice if the error from the server was a little more explicit.

File: ui-v2/app/routes/dc/acls/index.js (hunk 1 above)
Comment (johncowen) [reply ui-v2/app/routes/dc/acls/index.js]: Ah! hehe yeah sorted!

File: ui-v2/app/mixins/acl/with-actions.js [hunk 3]
```diff
@@ -23,6 +23,19 @@ export default Mixin.create(WithBlockingActions, {
...
     },
+    // TODO: This is also used in tokens, probably an opportunity to dry this out
+    logout: function(item) {
+      return get(this, 'feedback').execute(() => {
+        return get(this, 'settings')
+          .delete('token')
+          .then(() => {
+            // in this case we don't do the same as delete as we want to go to the new
+            // dc.acls.tokens page. If we get there via the dc.acls redirect/rewrite
+            // the we lose the flash message
```
Comment (DingoEatingFuzz) [inline ui-v2/app/mixins/acl/with-actions.js]: s/the/then

Comment (johncowen) [reply ui-v2/app/mixins/acl/with-actions.js]: ah doh! Sorted!

File: ui-v2/app/services/repository/token.js (hunk 2 above)
Comment (johncowen) [reply ui-v2/app/services/repository/token.js]: Hmm yeah it's not great, the entire message is actually:

`rpc error making call: rpc: can't find method ACL.TokenList`

or

`rpc error making call: rpc: can't find method ACL.TokenRead`

I'm gonna change what I'm searching for to:

`rpc error making call: rpc: can't find method ACL`

I think, so add `ACL` just incase

File: ui-v2/app/services/repository/token.js [hunk 4]
```diff
@@ -6,6 +6,7 @@ import { Promise } from 'rsvp';
 import statusFactory from 'consul-ui/utils/acls-status';
 const status = statusFactory(Promise);
 const MODEL_NAME = 'token';
+const UNKNOWN_METHOD_ERROR = "rpc error making call: rpc: can't find method";
```
Comment (rboyer) [inline ui-v2/app/services/repository/token.js]: If you wanted to be slightly more specific about it:
```
rpc error making call: rpc: can't find method ACL
```
would catch just ACL missing method errors (likely due to aclv2 upgrade)

File: ui-v2/app/templates/dc/acls/-form.hbs [hunk 5]
```diff
@@ -14,7 +14,7 @@
         </div>
         <label class="type-text">
             <span>Policy <a href="{{env 'CONSUL_DOCUMENTATION_URL'}}/guides/acl.html#rule-specification" rel="help noopener noreferrer" target="_blank">(HCL Format)</a></span>
-            {{code-editor class=(if item.error.Rules 'error') name='Rules' value=item.Rules onkeyup=(action 'change')}}
+            {{code-editor class=(if item.error.Rules 'error') name='Rules' value=item.Rules syntax="hcl" onkeyup=(action 'change')}}
```
Comment (rboyer) [inline ui-v2/app/templates/dc/acls/-form.hbs]: Did you mean to include this unrelated change?

File: ui-v2/app/templates/dc/acls/tokens/-fieldsets-legacy.hbs [hunk 6]
```diff
@@ -15,7 +15,7 @@
 {{/if}}
         <label class="type-text">
             <span>Rules <a href="{{env 'CONSUL_DOCUMENTATION_URL'}}/guides/acl.html#rule-specification" rel="help noopener noreferrer" target="_blank">(HCL Format)</a></span>
-            {{code-editor class=(if item.error.Rules 'error') name='Rules' value=item.Rules onkeyup=(action 'change' 'Rules')}}
+            {{code-editor class=(if item.error.Rules 'error') name='Rules' syntax='hcl' value=item.Rules onkeyup=(action 'change' 'Rules')}}
```
Comment (rboyer) [inline ui-v2/app/templates/dc/acls/tokens/-fieldsets-legacy.hbs]: Did you mean to include this?

File: ui-v2/app/utils/acls-status.js [hunk 7]
```diff
@@ -3,6 +3,10 @@
 // has a valid token
 // Right now this is very acl specific, but is likely to be
 // made a bit more less specific
+
+// This is repeated from repository/token, I'd rather it was repeated and not imported
+// for the moment at least
+const UNKNOWN_METHOD_ERROR = "rpc error making call: rpc: can't find method";
```
Comment (rboyer) [inline ui-v2/app/utils/acls-status.js]: reminder when you edit this to include `ACL` edit all copies

File: ui-v2/app/utils/acls-status.js [hunk 8]
```diff
@@ -23,6 +27,12 @@ export default function(P = Promise) {
       [propName]: p
         .catch(function(e) {
           switch (e.errors[0].status) {
+            case '500':
+              if (e.errors[0].detail.indexOf(UNKNOWN_METHOD_ERROR) === 0) {
```
Comment (rboyer) [inline ui-v2/app/utils/acls-status.js]: i think you want `!== -1` so that middle-of-the-string matches work.  When proxying between datacenters you will accumulate extra `rpc error making call:` prefixes for each hop

File: ui-v2/app/services/repository/token.js (hunk 2 above)
Comment (rboyer) [reply ui-v2/app/services/repository/token.js]: i think you want `!== -1` so that middle-of-the-string matches work.  When proxying between datacenters you will accumulate extra `rpc error making call:` prefixes for each hop

Comment (johncowen) [reply ui-v2/app/services/repository/token.js]: So what will that error look like?

`rpc error making call: rpc error making call: rpc error making call: rpc: can't find method ACL.TokenRead`

?

File: ui-v2/app/templates/dc/acls/-form.hbs (hunk 5 above)
Comment (johncowen) [reply ui-v2/app/templates/dc/acls/-form.hbs]: Yes, definite decision to include it here, checked this with jack, there are going to be a few more unrelated changes to come on this also I think.
